### ISSUE TITLE
onRetry doesn't work with async function

### ISSUE THREAD

**Issue by deefactorial** [jonbern/fetch-retry#29]
Use case is to get the response body json in the onRetry function;
```
retryOn: async (attempt: number, error: Error, response: Response) => false
```
will retry forever because it returns a promise and if (Promise) is true

so I can't do something like:
```
retryOn: async (attempt: number, error: Error, response: Response) => {
    return await response.json().property === true
}
```

**Comment by jonbern** [comment]
Thank you for creating an issue and pull request. Apologies for my late response.

I'm not 100% convinced this is a common usage pattern, and I'm hesitant of adding complexity to the code which isn't strictly needed, as this increases the chance of defects in the future. One of the original goals of this library was to keep it simple, and I want to try to keep it that way.

I will keep your issue open and if there's a lot of support for it going forward, I will reconsider my position.

**Comment by sangeetha5491** [comment]
https://github.com/jonbern/fetch-retry/pull/20 needs to work for the latest version as well. retries and retryOn does not work together. would be really useful if they did. Please consider fixing it again

**Comment by sangeetha5491** [comment]
can i raise a PR again for this issue?

**Comment by hoffination** [comment]
I also have the need for something like this. When running some E2E tests I need to sign a user up for a new account. They get an email confirmation number when signing up.

I'd like to be able to check the content of the result to see how many emails the user has received. Then use a retry with a falloff a limited number of times.

**Comment by jonbern** [comment]
The way I understand your use-case is that you need to make a network call to determine the number of e-mail a user has received, and based on that, determine whether or not to check again, and eventually fail the test if the user has not received x amount of emails.

In that case, I would say it's a bit outside of the scope of `fetch-retry`, because it primarily targets network issues, and potentially transient HTTP statuses.

Would it be possible to rewrite your E2E test to run in a `for` loop where you check the number of emails the user has received, apply the fall-of delay using a combination of a promise and `setTimeout` to wait, and then determine whether to try again? The `for` loop would iterate the max number of times you want to recheck the e-mail received status.

**Comment by rangle-benhofferber** [comment]
Yep, you understand correctly my use case and I see the elegance in the simplicity of what you're trying to accomplish. I wouldn't want to complicate things needlessly.

I'd also like to bring up another concern with my client's application that may be related to "potentially transient HTTP statuses." While it's not pretty or RESTful, their Backend-For-Frontend (BFF) service assumes ownership of the default error codes and passes along the status of the underlying micro-services in the body of the response.

Responses can look like `{ "status": "ok" }` or `{ "status": "nok", "error": 404, "message": "..." }` while both have the status of 200.

This makes it unfortunately a requirement of some of these tests to evaluate the underlying status in the body to get the true status.

**Comment by hoffination** [comment]
Yep, you understand correctly my use case and I see the elegance in the simplicity of what you're trying to accomplish. I wouldn't want to complicate things needlessly.

I'd also like to bring up another concern with my client's application that may be related to "potentially transient HTTP statuses." While it's not pretty or RESTful, their Backend-For-Frontend (BFF) service assumes ownership of the default error codes and passes along the status of the underlying micro-services in the body of the response.

Responses can look like { "status": "ok" } or { "status": "nok", "error": 404, "message": "..." } while both have the status of 200.

This makes it unfortunately a requirement of some of these tests to evaluate the underlying status in the body to get the true status.

**Comment by jonbern** [comment]
I agree, it doesn't sound very `RESTful`, and I believe that is the root cause of the problems you are describing :-)

You could implement a utility to parse the non-standard responses, but I think it will only take you so far, because the underlying problem is still there.

**Comment by colemars** [comment]
Hm shouldn't a custom onRetry be at least able to reproduce the default behavior, though?

**Comment by x3cion** [comment]
Hey there, this has turned into an issue for us and I provided #51 as a possible solution that builds on #30.

We want the retryOn to not only be based on a status, but also on a possible error payload. since the response.json() function is async, the whole retryOn must be async to work that way.

Is this enough of a use case to consider the async retryOn PR?

**Comment by jonbern** [comment]
Thank you for providing the pull request. I think this proves that there is sufficient evidence that this is a feature worth having. The PR looks mostly good, but the TS definitions file needs to be updated as well now that `retryOn` can be an async function.

**Comment by x3cion** [comment]
As mentioned in #51, I added the missing definition change now. Is there anything else missing?

**Comment by jonbern** [comment]
Looks great now. Many thanks. I have merged the PR and cut a new version `4.1.0` with your changes.
